File: Source/Aladdin.CAPI.ANSI.PKCS11/Cipher/RC5_CBC_PAD.cs
﻿using System;
using Aladdin.PKCS11; 

namespace Aladdin.CAPI.ANSI.PKCS11.Cipher
{
	///////////////////////////////////////////////////////////////////////////////
	// Алгоритм шифрования RC5 в режиме CBC с дополнением PKCS5
	///////////////////////////////////////////////////////////////////////////////
	public class RC5_CBC_PAD : CAPI.PKCS11.BlockMode
	{
		// размер ключей, число раундов и параметры режима
		private int[] keySizes; private int rounds; private CipherMode.CBC parameters; 

		// конструктор
		public RC5_CBC_PAD(CAPI.PKCS11.Applet applet, int rounds, 
			
			// сохранить переданные параметры
			int[] keySizes, CipherMode.CBC parameters) : base(applet, PaddingMode.PKCS5)
		{ 
			// сохранить переданные параметры
			this.parameters = parameters; this.rounds = rounds; 

            // указать допустимые размеры ключей
            if (keySizes != null) this.keySizes = keySizes; 
            else {
                // получить информацию алгоритма
                MechanismInfo info = applet.GetAlgorithmInfo(API.CKM_RC5_CBC_PAD); 
            
                // указать допустимые размеры ключей
                this.keySizes = CAPI.KeySizes.Range(info.MinKeySize, info.MaxKeySize); 
            }
		}
		// параметры алгоритма
		public override Mechanism GetParameters(CAPI.PKCS11.Session sesssion)
		{
            // указать параметры алгоритма
            Parameters.CK_RC5_CBC_PARAMS rc5Parameters = 
                new Parameters.CK_RC5_CBC_PARAMS(
                    parameters.BlockSize / 2, rounds, parameters.IV
            ); 
		    // вернуть параметры алгоритма
		    return new Mechanism(API.CKM_RC5_CBC_PAD, rc5Parameters); 
		}
        // тип ключей
        public override SecretKeyFactory KeyFactory { get { return Keys.RC5.Instance; }}
		// размер ключа в байтах
		public override int[] KeySizes { get { return keySizes; }}
		// размер блока
		public override int BlockSize { get { return parameters.BlockSize; }} 

		// режим алгоритма
		public override CipherMode Mode { get { return parameters; }}
	} 
}

File: Source/Aladdin.CAPI.ANSI.PKCS11.Java/src/aladdin/capi/ansi/pkcs11/cipher/RC5_CBC_PAD.java
package aladdin.capi.ansi.pkcs11.cipher;
import aladdin.pkcs11.*; 
import aladdin.pkcs11.jni.*; 
import aladdin.capi.*; 
import aladdin.capi.pkcs11.*; 
import aladdin.capi.ansi.keys.*;
import java.io.*;

///////////////////////////////////////////////////////////////////////////////
// Алгоритм шифрования RC5 в режиме CBC с дополнением PKCS5
///////////////////////////////////////////////////////////////////////////////
public class RC5_CBC_PAD extends aladdin.capi.pkcs11.BlockMode
{
	// размер ключей и число раундов
	private final int[] keySizes; private final int rounds; 
    // параметры режима
    private final CipherMode.CBC parameters; 

	// конструктор
	public RC5_CBC_PAD(Applet applet, int rounds, 
        int[] keySizes, CipherMode.CBC parameters) throws IOException
	{ 
		// сохранить переданные параметры
		super(applet, PaddingMode.PKCS5); this.rounds = rounds; this.parameters = parameters;
        
        // указать допустимые размеры ключей
        if (keySizes != null) this.keySizes = keySizes; 
        else {
            // получить информацию алгоритма
            MechanismInfo info = applet.getAlgorithmInfo(API.CKM_RC5_CBC_PAD); 
            
            // указать допустимые размеры ключей
            this.keySizes = KeySizes.range(info.minKeySize(), info.maxKeySize()); 
        }
	}
	// параметры алгоритма
	@Override public Mechanism getParameters(Session sesssion)
	{
        // указать параметры алгоритма
        CK_RC5_CBC_PARAMS rc5Parameters = new CK_RC5_CBC_PARAMS(
            parameters.blockSize() / 2, rounds, parameters.iv()
        ); 
        // вернуть параметры алгоритма
        return new Mechanism(API.CKM_RC5_CBC_PAD, rc5Parameters); 
	}
    // тип ключа
    @Override public final SecretKeyFactory keyFactory() { return new RC5(keySizes); } 
	// размер блока
	@Override public final int blockSize() { return parameters.blockSize(); } 
    
	// режим алгоритма
	@Override public CipherMode mode() { return parameters; }
} 
